File: lib/hubclient/transports/mqtttransport/MqttTransport.go
package mqtttransport

import (
	"context"
	"crypto/x509"
	"encoding/json"
	"errors"
	"fmt"
	"github.com/eclipse/paho.golang/packets"
	"github.com/eclipse/paho.golang/paho"
	"github.com/hiveot/hub/lib/hubclient/transports"
	"github.com/hiveot/hub/lib/keys"
	"github.com/hiveot/hub/lib/ser"
	"github.com/hiveot/hub/lib/tlsclient"
	"log/slog"
	"net"
	"os"
	"strings"
	"time"
)

const keepAliveInterval = 30 // seconds
const reconnectDelay = 10 * time.Second
const withDebug = false
const withQos = 1
const MqttInMemUDSProd = "@/MqttInMemUDSProd" // production UDS name
const MqttInMemUDSTest = "@/MqttInMemUDSTest" // test server UDS name

// MqttHubTransport manages the hub server connection with hub event and action messaging
// This implements the IHubClient interface.
// This implementation is based on the Mqtt messaging system.
type MqttHubTransport struct {
	serverURL      string
	clientID       string
	caCert         *x509.Certificate
	conn           net.Conn
	pcl            *paho.Client
	requestHandler *Handler      // PahoRPC request handler
	timeout        time.Duration // request timeout
}

// AddressTokens returns the address separator and wildcards
func (hc *MqttHubTransport) AddressTokens() (sep string, wc string, rem string) {
	return "/", "+", "#"
}

// ConnectWithConn connects to a mqtt broker using the pre-established network connection
// This allows creating connections with any available means including tcp/tls/wss/uds/pipes
//
//	credentials is either the password or a signed JWT token
//	conn network connection to utilize
func (ht *MqttHubTransport) ConnectWithConn(credentials string, conn net.Conn) error {
	ctx := context.Background()

	// clients must use a unique connection ID otherwise the previous connection will be dropped
	hostName, _ := os.Hostname()
	connectID := fmt.Sprintf("%s-%s-%s", ht.clientID, hostName, time.Now().Format("20060102150405.000"))
	slog.Info("ConnectWithConn", "clientID", ht.clientID, "RemoteAddr", conn.RemoteAddr(), "connectID", connectID)

	// checks
	if ht.clientID == "" {
		err := fmt.Errorf("connect - Missing Login ID")
		return err
	} else if conn == nil {
		err := fmt.Errorf("connect - missing connection")
		return err
	}

	safeConn := packets.NewThreadSafeConn(conn)
	pahoCfg := paho.ClientConfig{
		ClientID:      ht.clientID,
		Conn:          safeConn,
		PacketTimeout: ht.timeout,
	}
	pahoCfg.OnClientError = func(err error) {
		// connection closing can cause this error.
		slog.Debug("OnClientError - connection closing",
			slog.String("err", err.Error()),
			slog.String("clientID", pahoCfg.ClientID))
	}
	pahoCfg.OnServerDisconnect = func(d *paho.Disconnect) {
		slog.Warn("ConnectWithNC:OnServerDisconnect: Disconnected from broker",
			"code", d.ReasonCode,
			"loginID", ht.clientID)
	}
	pcl := paho.NewClient(pahoCfg)
	cp := &paho.Connect{
		Password: []byte(credentials),
		Username: ht.clientID,
		ClientID: connectID,
		// TODO: consider including a signed nonce when connecting with key
		Properties:   &paho.ConnectProperties{},
		KeepAlive:    60,
		CleanStart:   true,
		UsernameFlag: true,
		PasswordFlag: credentials != "",
	}
	ctx, cancelFn := context.WithTimeout(context.Background(), ht.timeout)
	defer cancelFn()
	connAck, err := pcl.Connect(ctx, cp)
	_ = connAck

	if err != nil {
		var info string
		if connAck != nil {
			info = fmt.Sprintf("code %d, reason: '%s'", connAck.ReasonCode, connAck.Properties.ReasonString)
		}
		err = fmt.Errorf("%w %s", err, info)
		//err = fmt.Errorf("failed to connect. Reason: %d - %s",
		//	connAck.ReasonCode, connAck.Properties.ReasonString)
		return err
	}
	ht.conn = conn
	ht.pcl = pcl

	// last, create a request handler
	ht.requestHandler, err = NewHandler(ctx, ht.pcl)

	return err
}

// ConnectWithCert connects to the Hub server using TLS client certificate.
//
//	brokerURL of the server.
//	clientID to connect as
//	clientCert for certificate based authentication
//	caCert of the server
//func (hc *MqttHubTransport) ConnectWithCert(clientCert tls.Certificate) error {
//
//	conn, err := tlsclient.ConnectTLS(hc.serverURL, &clientCert, hc.caCert)
//	if err != nil {
//		return err
//	}
//	err = hc.ConnectWithConn("", conn)
//	return err
//}

// ConnectWithPassword connects to the Hub server using a login ID and password.
func (ht *MqttHubTransport) ConnectWithPassword(password string) error {

	conn, err := tlsclient.ConnectTLS(ht.serverURL, nil, ht.caCert)
	if err != nil {
		return err
	}
	err = ht.ConnectWithConn(password, conn)
	return err
}

// ConnectWithToken connects to the Hub server using a user JWT credentials secret
// The token clientID must match that of the client
// A private key might be required in future.
// This supports UDS connections with @/path or unix://@/path
//
// TODO: encrypt token with server public key so a MIM won't be able to get the token
//
//	kp is the key-pair of this client
//	jwtToken is the token obtained with login or refresh.
func (ht *MqttHubTransport) ConnectWithToken(kp keys.IHiveKey, jwtToken string) error {
	var conn net.Conn
	var err error
	if strings.HasPrefix(ht.serverURL, "unix://") {
		// mqtt server UDS listener doesn't use TLS
		// url.Parse doesn't recognize the @ in the path
		addr := ht.serverURL[7:]
		conn, err = net.Dial("unix", addr)
	} else if strings.HasPrefix(ht.serverURL, "@") {
		// using a UDS in-memory path without scheme
		conn, err = net.Dial("unix", ht.serverURL)
	} else {
		conn, err = tlsclient.ConnectTLS(ht.serverURL, nil, ht.caCert)
	}
	if err == nil {
		// clientID from the token is used
		err = ht.ConnectWithConn(jwtToken, conn)
	}
	if err != nil {
		err = fmt.Errorf("ConnectWithToken failed: %w", err)
	}
	return err
}

// CreateKeyPair returns a new set of serialized public/private key pair
func (ht *MqttHubTransport) CreateKeyPair() (cryptoKeys keys.IHiveKey) {
	k := keys.NewKey(keys.KeyTypeECDSA)
	return k
}

// Disconnect from the MQTT broker and unsubscribe from all topics and set
// device state to disconnected
func (ht *MqttHubTransport) Disconnect() {
	if ht.pcl != nil {
		time.Sleep(time.Millisecond * 10) // Disconnect doesn't seem to wait for all messages. A small delay ahead helps
		_ = ht.pcl.Disconnect(&paho.Disconnect{ReasonCode: 0})
		//time.Sleep(time.Millisecond * 10) // Disconnect doesn't seem to wait for all messages. A small delay ahead helps
		//ht.pcl = nil

		//ht.subscriptions = nil
		//close(ht.messageChannel)     // end the message handler loop
	}
}

// ParseResponse helper message to parse response and check for errors
func (ht *MqttHubTransport) ParseResponse(data []byte, resp interface{}) error {
	var err error
	if data == nil || len(data) == 0 {
		if resp != nil {
			err = fmt.Errorf("ParseResponse: client '%s', expected a response but none received", ht.clientID)
		} else {
			err = nil // all good
		}
	} else {
		if resp == nil {
			err = fmt.Errorf("ParseResponse: client '%s', received response but none was expected. data=%s",
				ht.clientID, data)
		} else {
			err = ser.Unmarshal(data, resp)
		}
	}
	return err
}

// Pub publishes a message and returns
func (ht *MqttHubTransport) Pub(topic string, payload []byte) (err error) {
	slog.Debug("Pub", "topic", topic)
	ctx, cancelFn := context.WithTimeout(context.Background(), ht.timeout)
	defer cancelFn()
	pubMsg := &paho.Publish{
		QoS:     withQos,
		Retain:  false,
		Topic:   topic,
		Payload: payload,
	}
	respMsg, err := ht.pcl.Publish(ctx, pubMsg)
	_ = respMsg
	if err != nil {
		return err
	}
	return err
}

// PubRequest publishes a request message and waits for an answer or until timeout
func (ht *MqttHubTransport) PubRequest(topic string, payload []byte) (resp []byte, err error) {
	slog.Info("PubRequest", "topic", topic)

	ctx, cancelFn := context.WithTimeout(context.Background(), ht.timeout)
	defer cancelFn()

	pubMsg := &paho.Publish{
		QoS:     withQos,
		Retain:  false,
		Topic:   topic,
		Payload: payload,
		Properties: &paho.PublishProperties{
			ContentType: "json",
			User: paho.UserProperties{{
				Key:   "test",
				Value: "test",
			}},
		},
	}
	// use the inbox as the custom response for this client instance
	// clone of rpc.go to workaround hangup when no response is received #111
	respMsg, err := ht.requestHandler.Request(ctx, pubMsg)
	//ar.Duration = time.Now().Sub(t1)
	if err != nil {
		return nil, err
	}

	// test alternative to handling errors since User properties aren't
	// passed through for some reason.
	if respMsg.Properties.ContentType == "error" {
		err = errors.New(string(respMsg.Payload))
		return nil, err
	}

	slog.Debug("PubRequest end:",
		slog.String("topic", topic),
		slog.String("ContentType (if any)", respMsg.Properties.ContentType),
	)
	return respMsg.Payload, err
}

// sendReply sends a reply on the response topic of the request
// This uses the same QoS as the request, without retain.
//
//	req is the request to reply to
//	optionally include a payload in the reply
//	optionally include an error message in the reply
func (ht *MqttHubTransport) sendReply(req *paho.Publish, payload []byte, errResp error) error {

	slog.Debug("sendReply", "topic", req.Topic, "responseTopic", req.Properties.ResponseTopic)

	responseTopic := req.Properties.ResponseTopic
	if responseTopic == "" {
		err2 := fmt.Errorf("sendReply. No response topic. Not sending reply.")
		slog.Error(err2.Error())
	}
	replyMsg := &paho.Publish{
		QoS:    req.QoS,
		Retain: false,
		Topic:  responseTopic,
		Properties: &paho.PublishProperties{
			CorrelationData: req.Properties.CorrelationData,
			User:            req.Properties.User,
			PayloadFormat:   req.Properties.PayloadFormat,
			ContentType:     req.Properties.ContentType,
		},
		Payload: payload,
	}
	if errResp != nil {
		replyMsg.Properties.ContentType = "error" // payload is an error message
		replyMsg.Properties.User.Add("error", errResp.Error())
		// for testing, somehow properties.user is not transferred
		replyMsg.Payload = []byte(errResp.Error())
	}
	ctx, cancelFn := context.WithTimeout(context.Background(), time.Second)
	defer cancelFn()
	_, err := ht.pcl.Publish(ctx, replyMsg)
	if err != nil {
		slog.Warn("sendReply. Error publishing response",
			slog.String("err", err.Error()))
	}
	return err
}

// Sub subscribes to a topic
func (ht *MqttHubTransport) Sub(topic string, cb func(topic string, msg []byte)) (transports.ISubscription, error) {
	slog.Info("Sub", "topic", topic)
	spacket := &paho.Subscribe{
		Properties: nil,
		Subscriptions: map[string]paho.SubscribeOptions{
			topic: {QoS: withQos},
		},
	}
	suback, err := ht.pcl.Subscribe(context.Background(), spacket)
	ht.pcl.Router.RegisterHandler(topic, func(m *paho.Publish) {
		slog.Info("Sub, received Msg:", "topic", m.Topic)
		//clientID := m.Properties.User.Get("clientID") // experimental

		// run this in the background to allow for reentrancy
		go func() {
			cb(m.Topic, m.Payload)
		}()
	})
	_ = suback
	hcSub := &PahoSubscription{
		topic:    topic,
		pcl:      ht.pcl,
		clientID: ht.clientID,
	}
	return hcSub, err
}

// SubRequest subscribes to a requests and sends a response
// Intended for actions, config and rpc requests
func (ht *MqttHubTransport) SubRequest(
	topic string, cb func(topic string, payload []byte) (reply []byte, err error)) (
	transports.ISubscription, error) {

	spacket := &paho.Subscribe{
		Properties: nil,
		Subscriptions: map[string]paho.SubscribeOptions{
			topic: {QoS: withQos},
		},
	}
	suback, err := ht.pcl.Subscribe(context.Background(), spacket)
	_ = suback
	ht.pcl.Router.RegisterHandler(topic, func(m *paho.Publish) {

		// run this in the background to allow for reentrancy
		go func() {
			reply, err := cb(m.Topic, m.Payload)

			// TODO: cleanup. Is there something useful in properties.user?
			//clientID:= m.Properties.User.Get("clientID")
			propsJson, _ := json.Marshal(m.Properties.User)
			slog.Debug("SubRequest. Properties.User",
				"props", propsJson)

			if err != nil {
				slog.Warn("SubRequest: handle request failed",
					slog.String("err", err.Error()),
					slog.String("topic", topic))

				err = ht.sendReply(m, nil, err)
			} else {
				err = ht.sendReply(m, reply, err)
			}
			if err != nil {
				slog.Error("SubRequest. Sending reply failed", "err", err)
			}
		}()
	})

	hcSub := &PahoSubscription{
		topic: topic,
		pcl:   ht.pcl,
	}

	return hcSub, err
}

// NewMqttTransport creates a new instance of the mqtt client.
//
// fullURL is the url with schema. If omitted this uses the in-memory UDS address,
// which only works with ConnectWithToken.
//
//	fullURL of broker to connect to, starting with "tls://", "wss://", "unix://"
//	clientID to connect as
//	keyPair with previously saved serialized public/private key pair, or "" to create one
//	caCert of the server to validate the server or nil to not check the server cert
func NewMqttTransport(fullURL string, clientID string, caCert *x509.Certificate) *MqttHubTransport {
	if fullURL == "" {
		fullURL = "unix://" + MqttInMemUDSProd
	}
	hc := &MqttHubTransport{
		serverURL: fullURL,
		clientID:  clientID,
		caCert:    caCert,
		timeout:   time.Second * 10, // 10 for testing
	}
	return hc
}

File: libjs/src/hubclient/transports/mqtttransport/MqttTransport.ts

// MqttTransport
import { ConnInfo, ConnectionStatus } from "../IHubTransport";
import type { IHubTransport } from "../IHubTransport";
import * as mqtt from 'mqtt';
import * as os from "os";
import type { IHiveKey } from "@hivelib/keys/IHiveKey";
import { ECDSAKey } from "@hivelib/keys/ECDSAKey";
import type { IPublishPacket } from 'mqtt';
import type { QoS } from "mqtt-packet";
import * as tslog from 'tslog';

const log = new tslog.Logger()

export class MqttTransport implements IHubTransport {
    // expect mqtt://addr:port/
    fullURL: string
    clientID: string
    caCertPem: string
    instanceID: string = ""
    inboxTopic: string = ""
    timeoutMSec: number = 30000 // request timeout in millisec
    qos: QoS = 0

    // application handler of connection status change
    connectHandler: null | ((status: ConnectionStatus, info: string) => void) = null;
    // application handler of incoming messages
    eventHandler: null | ((topic: string, payload: string) => void) = null;
    // application handler of incoming request-response messages
    requestHandler: null | ((topic: string, payload: string) => string) = null;

    // myKeys:IHiveKey
    // https://github.com/mqttjs/MQTT.js/
    mcl: mqtt.MqttClient | null = null

    // map of correlationID to handler receiving a reply or timeout error
    replyHandlers: { [index: string]: (corrID: string, reply: string) => void };

    constructor(
        fullURL: string, clientID: string, caCertPem: string) {

        this.fullURL = fullURL
        this.clientID = clientID
        this.caCertPem = caCertPem
        this.replyHandlers = {}
    }

    public addressTokens(): { sep: string, wc: string, rem: string } {
        return { sep: "/", wc: "+", rem: "#" }
    }

    // connect and subscribe to the inbox
    public async connectWithPassword(password: string): Promise<void> {
        // let urlParts = new URL(this.fullURL)
        let p = new Promise<void>((resolve, reject) => {

            // log.info("connectWithPassword; url:", this.fullURL, "; clientID:", this.clientID)

            let timestamp = Date.now().toString() // msec since epoch
            let rn = Math.random().toString(16).substring(2, 8)
            this.instanceID = this.clientID + "-" + os.hostname + "-" + timestamp + "-" + rn
            // see https://github.com/mqttjs/MQTT.js/ for options
            let opts: mqtt.IClientOptions = {
                username: this.clientID,
                password: password,
                //
                ca: this.caCertPem,
                clean: true, // false
                clientId: this.instanceID,
                keepalive: 60,
                protocolVersion: 5, // MQTT 5 is required for rpc
                rejectUnauthorized: (this.caCertPem != ""),
                // properties: {
                //     userProperties: {
                //         "clientID": this.clientID
                //     }
                // }

            }

            this.mcl = mqtt.connect(this.fullURL, opts)

            this.mcl.on("connect", (packet: any) => {
                log.info("MQTT client connected to:", this.fullURL)
                // subscribe to inbox; server only allows inbox subscription with the clientID
                // to prevent subscribing to other client's inboxes.
                this.inboxTopic = "_INBOX/" + this.instanceID
                if (this.mcl) {
                    let inboxSub = this.mcl.subscribe(this.inboxTopic)
                    log.info("subscribed to ", this.inboxTopic)
                }
                resolve()
                if (this.connectHandler) {
                    this.connectHandler(ConnectionStatus.Connected, ConnInfo.Success)
                }
            })
            this.mcl.on("disconnect", (packet: mqtt.IDisconnectPacket) => {
                log.info("MQTT server disconnected:")
                let reason: string = packet.reasonCode?.toString() || ""
                let err: Error | null = null
                if (packet.reasonCode != 0) {
                    err = new Error(reason + ":" + packet.properties?.reasonString)
                }
                if (this.connectHandler) {
                    this.connectHandler(ConnectionStatus.Disconnected, err?.message || "")
                }
            })

            this.mcl.on("error", (err: Error) => {
                log.error("MQTT error: ", err.message)
                reject(err)
            })
            this.mcl.on("message",
                (topic: string, payload: Buffer, packet: IPublishPacket) => {
                    // log.info("on message. topic:", topic)
                    this.onRawMessage(topic, payload, packet)
                })
            this.mcl.on("offline", () => {
                log.error("Connection lost")
            })
            // this.mcl.on("packetreceive", (packet: mqtt.Packet) => {
            //     log.error("packetReceive: cmd:", packet.cmd)
            // })


        })
        return p
    }

    public async connectWithToken(key: IHiveKey, token: string): Promise<void> {
        // TODO: encrypt token with server public key so a MIM won't be able to get the token
        return this.connectWithPassword(token)
    }

    // this mqtt transport uses ECDSA keys
    public createKeyPair(): IHiveKey {
        let kp = new ECDSAKey().initialize()
        return kp
    }

    public disconnect(): void {
        if (this.mcl != null) {
            this.mcl.end()
            this.mcl = null
        }
    }

    // handle incoming mqtt message.
    // This determines the type of message and passes it on to the corresponding handlers.
    //  * INBOX message: The handler registered to the correlationID will be invoked.
    //  * Message with replyTo address: the RequestHandler will be invoked and its result is sent to the replyTo address.
    //  * Message with no replyTo address: The MessageHandler will be invoked.
    onRawMessage(topic: string, payload: Buffer, packet: mqtt.IPublishPacket): void {
        // intercept INBOX replies
        let payloadStr = payload.toString()
        let cordata = packet.properties?.correlationData
        // TODO: an inbox address could include the correlation ID. Is this worth it for mqtt-v3?
        let corid = cordata?.toString() || ""
        try {
            if (topic.startsWith(this.inboxTopic)) {
                // this is a response
                if (corid == "") {
                    log.error("onRawMessage: ignored response without correlation ID.")
                } else {
                    // find the handler for the correlation ID.
                    let handler = this.replyHandlers[corid]
                    handler(corid, payloadStr)
                    return
                }
            } else if (packet.properties?.responseTopic) {
                let err: Error | null = null
                let reply: string = ""
                // this is a request message that asks for a response
                if (!corid) {
                    err = Error("onRawMessage: request without correlationID. Topic:" + topic)
                } else if (!this.requestHandler) {
                    err = Error("onRawMessage: No handler for topic:" + topic)
                } else {
                    // this is a request that asks for a reponse
                    try {
                        reply = this.requestHandler(topic, payloadStr)
                    } catch (e: any) {
                        // catch this error so a reply can be sent
                        err = Error("onRawMessage ERROR: exception:", e)
                    }
                }
                this.pubReply(reply, err, packet)

            } else {
                // this is an event type message
                if (this.eventHandler) {
                    this.eventHandler(topic, payloadStr)
                }
            }
        } catch (e) {
            log.error("onRawMessage: Exception handling message. topic:", topic, ", err:", e)
        }
    }

    // send an event and return immediately
    public async pubEvent(address: string, payload: string): Promise<void> {
        if (this.mcl) {

            let opts: mqtt.IClientPublishOptions = {
                qos: this.qos,
            }
            // this.mcl.publish(address, payload)
            this.mcl.publish(address, payload, opts)
        }
        return
    }

    // send an action or RPC request and wait for a reply
    public async pubRequest(address: string, payload: string): Promise<string> {

        let p = new Promise<string>((resolve, reject) => {
            let rn = Math.random().toString(16).substring(2, 8)
            // let corid = Date.now().toString(16) + "." + rn
            let rightnow = new Date()
            let corid = rightnow.toISOString() + "." + rn
            let opts: mqtt.IClientPublishOptions = {
                qos: this.qos,
                properties: {
                    responseTopic: this.inboxTopic,
                    correlationData: Buffer.from(corid)
                }
            }
            this.mcl ? this.mcl.publish(
                address, payload, opts, (err) => {
                    if (err) {
                        log.info(err)
                        reject()
                        throw (err)
                    } else {
                        log.info('pubRequest: Request sent to ' + address + 'with correlationID:', corid)
                    }
                }) : "";
            // if the timer isn't cancelled in time, it will reject the request
            // this should work with both node and browser
            // FIXME remove reply handler after use
            let timoutID = setTimeout(() => reject("timeout"), this.timeoutMSec)
            this.replyHandlers[corid] =
                function (corrID: string, payload: string): void {
                    // received a reply, cancel the timer and resolve the request
                    clearTimeout(timoutID)
                    log.info("pubRequest: invoking reply handler")
                    resolve(payload)
                }
        })
        return p
    }

    // send a reply to a request
    async pubReply(payload: string, err: Error | null, request: IPublishPacket) {
        if ((!this.mcl) || (!request.properties?.correlationData) || !request.properties.responseTopic) {
            let err = Error("pubReply: missing replyTo or correlationData")
            log.error(err)
            throw err
        }
        let corid = request.properties?.correlationData
        // FIXME send as a reply without responseTopic
        let opts: mqtt.IClientPublishOptions = {
            qos: this.qos,
            properties: {
                correlationData: corid,
            }
        }
        // typescript doesn't recognize that opts.properties is already set
        if (err && (!!opts.properties)) {
            let userProp = { error: err.message };
            opts.properties.userProperties = userProp;
        }
        let replyTo = request.properties?.responseTopic
        this.mcl.publish(replyTo, payload, opts, (err) => {
            if (err) {
                // failed to send a reply
                log.error("pubReply: failed to send reply. err=", err)
                throw (err)
            } else {
                log.info('pubReply: Request sent with correlation ID:', corid.toString())
            }
        })
        return
    }

    // Set the callback of connect/disconnect updates
    public setConnectHandler(handler: (status: ConnectionStatus, info: string) => void): void {
        this.connectHandler = handler
    }

    // Set the handler of incoming messages
    public setEventHandler(handler: (topic: string, payload: string) => void): void {
        this.eventHandler = handler
    }

    // Set the handler of incoming requests-response calls.
    // The result of the handler is sent as a reply.
    // Intended for handling actions and RPC requests.
    public setRequestHandler(handler: (topic: string, payload: string) => string): void {
        this.requestHandler = handler
    }

    // subscribe to a topic
    public async subscribe(topic: string): Promise<void> {
        let p = new Promise<void>((resolve, reject) => {
            if (!this.mcl) {
                throw ("subscribe: no server connection");
            }
            // this.mcl.subscribe(topic, opts, (err, granted) => {
            let opts: mqtt.IClientSubscribeOptions = {
                qos: this.qos
            }
            this.mcl.subscribe(topic, opts, (err, granted) => {
                // remove registration if subscription fails
                if (err) {
                    log.error("subscribe: failed: " + err);
                    reject(err)
                } else {       // all good
                    log.info("subscribe: topic:" + topic);
                    resolve()
                }
            })
        })
        return p;

    }

    public unsubscribe(address: string) {
        if (this.mcl) {
            this.mcl.unsubscribe(address);
        }
    }
}

// Create a new MQTT transport using websockets over SSL
//
// fullURL schema supports: mqtt, mqtts, tcp, tls, ws, wss, wxs, alis
//
//
// @param fullURL is the websocket address: wss://address:port/path
// @param clientID is the client's connection ID
// @param caCertPem is the pem encoded CA certificate, if available. Use "".
// @param onMessage is the message handler for subscriptions
// export function NewMqttTransport(fullURL:string, clientID:string, caCertPem:string,
//                                  onMessage:(topic:string,msg:string)=>void): MqttTransport {
//     //
//     // caCertPool := x509.NewCertPool()
//     // if caCert != nil {
//     //     caCertPool.AddCert(caCert)
//     // }
//     // tlsConfig := &tls.Config{
//     //     RootCAs:            caCertPool,
//     //         InsecureSkipVerify: caCert == nil,
//     // }
//
//
//     let tp = new MqttTransport(fullURL,clientID,caCertPool,onMessage)
//     return tp
// }